### ISSUE TITLE
Plugin seems to be hanging in some circumstances

### ISSUE THREAD

**Issue by IRCody** [intelsdi-x/snap-plugin-collector-cpu#33]
This plugin seems to be hanging under certain conditions, which stops the workflow and causes some odd framework behavior. The framework behavior is tracked by [#1142](https://github.com/intelsdi-x/snap/issues/1142) on the snap repo.

The environment I tested on:
OS: ubuntu 14.04
Snap commit: bc78f4b57ca63cf00642ebf85e54d924bac3e80b, master at the time I tested.
cpu plugin commit: c4a90ddf785835a3d6d830eb0292aa418b2bcd9e, also master at the time tested.

The steps to reproduce:

0. Make snapd and cpu plugin
1. Start snapd, loading the mock plugins. In $SNAP_PATH/bin do `./snapd -t 0 -l 1 -a ../plugin`.
2. Load cpu plugin, `snapctl plugin load snap-plugin-collector-cpu`
3. Try to start the example task, `snapctl task create -t examples/task/cpu-file.json`. I modified the task slightly. The modifications include changing "file" to "mock-file" so that I can use the mock-file publisher and reducing the number of metrics to 1 to decrease the amount of noise in the logs.
4. Observe that task appears to be 'frozen'. Can be verified by checking snapds logs, and noting lack of any collection behavior, or by monitoring the output file specified in the task or by attempting to do sequential task lists which will show lack of increasing hits or misses or failure messages:
```
‹master*› % snapctl task list
ID                                       NAME                                            STATE           HIT     MISS    FAIL    CREATED                 LAST FAILURE
432a3e3d-5099-45d3-8568-7902266ac0b2     Task-432a3e3d-5099-45d3-8568-7902266ac0b2       Running         4       0       0       4:48PM 8-15-2016
‹master*› % snapctl task list
ID                                       NAME                                            STATE           HIT     MISS    FAIL    CREATED                 LAST FAILURE
432a3e3d-5099-45d3-8568-7902266ac0b2     Task-432a3e3d-5099-45d3-8568-7902266ac0b2       Running         4       0       0       4:48PM 8-15-2016
‹master*› % snapctl task list
ID                                       NAME                                            STATE           HIT     MISS    FAIL    CREATED                 LAST FAILURE
432a3e3d-5099-45d3-8568-7902266ac0b2     Task-432a3e3d-5099-45d3-8568-7902266ac0b2       Running         4       0       0       4:48PM 8-15-2016
```

For me it always seems to happen at the 4th collection. Collections 1-3 do succeed and give me the output I'd expect of:
```
2016-08-15 16:48:24.063897451 -0700 PDT|/intel/procfs/cpu/0/utilization_percentage|3167.200599906572|tags[plugin_running_on:wcp1]
2016-08-15 16:48:25.059392521 -0700 PDT|/intel/procfs/cpu/0/utilization_percentage|3167.200599906572|tags[plugin_running_on:wcp1]
2016-08-15 16:48:26.059262474 -0700 PDT|/intel/procfs/cpu/0/utilization_percentage|3167.200599906572|tags[plugin_running_on:wcp1]
```

The last few lines of logs from snapd:
```
DEBU[2016-08-15T16:48:26-07:00] starting collector job                        _module=scheduler-job block=run job-type=collector metric-count=1
DEBU[2016-08-15T16:48:26-07:00] cache miss [/intel/procfs/cpu/0/utilization_percentage:0]  _module=routing-cache hits=0 misses=3 namespace=/intel/procfs/cpu/0/utilization_percentage:0
DEBU[2016-08-15T16:48:26-07:00] plugin selected                               _module=control-routing block=select hitcount=2 index=collector:cpu:v3:id1 pool size=1 strategy=least-recently-used
DEBU[2016-08-15T16:48:26-07:00] collector run completed                       _module=scheduler-job block=run job-type=collector metric-count=1
DEBU[2016-08-15T16:48:26-07:00] Batch submission of process and publish nodes  _block=work-jobs _module=scheduler-workflow count-process-nodes=1 count-publish-nodes=0 parent-node-type=collector task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] Submitting process job                        _block=submit-process-job _module=scheduler-workflow parent-node-type=collector process-name=passthru process-version=-1 task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] starting processor job                        _module=scheduler-job block=run content-type=snap.gob job-type=processor plugin-config=map[] plugin-name=passthru plugin-version=-1
DEBU[2016-08-15T16:48:26-07:00] plugin selected                               _module=control-routing block=select hitcount=2 index=processor:passthru:v1:id1 pool size=1 strategy=least-recently-used
DEBU[2016-08-15T16:48:26-07:00] Process job completed                         _block=submit-process-job _module=scheduler-workflow parent-node-type=collector process-name=passthru process-version=-1 task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] Batch submission of process and publish nodes  _block=work-jobs _module=scheduler-workflow count-process-nodes=0 count-publish-nodes=1 parent-node-type=processor task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] Submitting publish job                        _block=submit-publish-job _module=scheduler-workflow parent-node-type=processor publish-name=mock-file publish-version=-1 task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] starting publisher job                        _module=scheduler-job block=run content-type=snap.gob job-type=publisher plugin-config=map[file:{Value:/tmp/published_cpu.log}] plugin-name=mock-file plugin-version=-1
DEBU[2016-08-15T16:48:26-07:00] plugin selected                               _module=control-routing block=select hitcount=2 index=publisher:mock-file:v3:id1 pool size=1 strategy=least-recently-used
DEBU[2016-08-15T16:48:26-07:00] Publish job completed                         _block=submit-publish-job _module=scheduler-workflow parent-node-type=processor publish-name=mock-file publish-version=-1 task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] Batch submission complete                     _block=work-jobs _module=scheduler-workflow count-process-nodes=0 count-publish-nodes=1 parent-node-type=processor task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] Batch submission complete                     _block=work-jobs _module=scheduler-workflow count-process-nodes=1 count-publish-nodes=0 parent-node-type=collector task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:26-07:00] task spin loop                                _module=scheduler-task
DEBU[2016-08-15T16:48:26-07:00] event received                                _block=handle-events _module=scheduler-events event-namespace=Scheduler.MetricsCollected metric-count=1 task-id=432a3e3d-5099-45d3-8568-7902266ac0b2
INFO[2016-08-15T16:48:26-07:00] API request                                   _module=_mgmt-rest index=3 method=GET url=/v1/tasks
INFO[2016-08-15T16:48:26-07:00] API response                                  _module=_mgmt-rest index=3 method=GET status=OK status-code=200 url=/v1/tasks
INFO[2016-08-15T16:48:27-07:00] Starting workflow for task (432a3e3d-5099-45d3-8568-7902266ac0b2\Task-432a3e3d-5099-45d3-8568-7902266ac0b2)  _block=workflow-start _module=scheduler-workflow task-id=432a3e3d-5099-45d3-8568-7902266ac0b2 task-name=Task-432a3e3d-5099-45d3-8568-7902266ac0b2
DEBU[2016-08-15T16:48:27-07:00] starting collector job                        _module=scheduler-job block=run job-type=collector metric-count=1
DEBU[2016-08-15T16:48:27-07:00] cache miss [/intel/procfs/cpu/0/utilization_percentage:0]  _module=routing-cache hits=0 misses=4 namespace=/intel/procfs/cpu/0/utilization_percentage:0
DEBU[2016-08-15T16:48:27-07:00] plugin selected                               _module=control-routing block=select hitcount=3 index=collector:cpu:v3:id1 pool size=1 strategy=least-recently-used
```

**Comment by katarzyna-z** [comment]
I'm not able to reproduce this issue using commits (for snap and plugin) which you mentioned on Ubuntu 16.04. I'll try to reproduce the issue on Ubuntu 14.04.

**Comment by IRCody** [comment]
Thanks for taking a look. I'm happy to answer any questions or do more specific testing on my end if needed. Just let me know.

**Comment by katarzyna-z** [comment]
I have been trying to reproduce the issue using commits (for snap and plugin) which you mentioned on this version of Ubuntu:
```
test@test-VirtualBox:~/snap/build/bin$ cat /etc/lsb-release ; uname -a
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.4 LTS"
Linux test-VirtualBox 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
with this task manifest:
```
{
    "version": 1,
    "schedule": {
        "type": "simple",
        "interval": "1s"
    },
    "workflow": {
        "collect": {
            "metrics": {
	"/intel/procfs/cpu/0/utilization_percentage" : {}
            },
            "process": [
                {
                    "plugin_name": "passthru",
                    "process": null,
                    "publish": [
                        {
                            "plugin_name": "mock-file",
                            "config": {
                                "file": "/tmp/published_cpu.log"
                            }
                        }
                    ],
                    "config": null
                }
            ],
            "publish": null
        }
    }
}
```

and I'm also not able to reproduce. 

Could you check whether you have the same version of ubuntu?

**Comment by IRCody** [comment]
Here is my output:

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.4 LTS"
Linux wcp1 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

I think my task manifest is identical to yours but pasted here for completeness:

```json
{
    "version": 1,
    "schedule": {
        "type": "simple",
        "interval": "1s"
    },
    "workflow": {
        "collect": {
            "metrics": {
        "/intel/procfs/cpu/0/utilization_percentage" : {}
            },
            "process": [
                {
                    "plugin_name": "passthru",
                    "process": null,
                    "publish": [
                        {
                            "plugin_name": "mock-file",
                            "config": {
                                "file": "/tmp/published_cpu.log"
                            }
                        }
                    ],
                    "config": null
                }
            ],
            "publish": null
        }
    }
}
```

I just ran through the sequence again and it has the same issue reliably.

**Comment by jcooklin** [comment]
@katarzyna-z: I was unable to reproduce in a 14.04 container or VM however when I run the task described above on a physical host with many cores (24) it locks up just as described above.

**Comment by IRCody** [comment]
My failing machine is also running significant # of cores if that helps narrow it down. I tried looking at the code but was unable to trace where that would be an issue. Happy to try any suggested debugging steps.

**Comment by katarzyna-z** [comment]
Maybe this plugin has a problem with reading large number of lines in the output of "/proc/stat". My output is quite short: 
```
cpu  1117969 9786 642692 75235064 78961 0 1985 0 0 0
cpu0 181668 1851 91026 9347467 1287 0 240 0 0 0
cpu1 178988 2047 92529 9345906 4403 0 102 0 0 0
cpu2 171898 1488 99974 9346585 2771 0 390 0 0 0
cpu3 175023 1201 128246 9317586 1680 0 544 0 0 0
cpu4 96413 969 49860 9434260 65728 0 136 0 0 0
cpu5 108160 649 57223 9487285 206 0 110 0 0 0
cpu6 99167 1074 95344 9456659 882 0 151 0 0 0
cpu7 106650 505 28488 9499313 2001 0 308 0 0 0
intr 84239747 17 2 0 0 0 0 0 8 1 4 0 0 4 0 0 0 29 0 0 0 0 0 0 33 0 0 0 0 0 0 0 0 0 0 0 0 793450 671853 0 1352574 1973130 1287 12 1425 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
ctxt 346141458
btime 1471432466
processes 70197
procs_running 1
procs_blocked 0
softirq 39833988 371 15734360 20174 3570085 675249 0 56492 13422891 0 6354366
```
I'll see.

**Comment by lynxbat** [comment]
@katarzyna-z I was thinking the same thing while reading this thread (scanning a large return). I know @jcooklin ran into interesting issues with scan on large strings.

**Comment by IRCody** [comment]
>Maybe this plugin has a problem with reading large number of lines in the output of "/proc/stat" or calculation of metrics takes too much time. My output of "/proc/stat" is quite short:

My output from `/proc/stat` is:
```
cpu  4544659691 116971 6042061976 44122699686 5625601812 0 640911165 0 211628 0
cpu0 123707048 7870 159639198 832897028 130366882 0 18076356 0 2586 0
cpu1 119830976 4742 158775324 838885414 131072654 0 17989525 0 2064 0
cpu2 119711957 3237 158799879 839291416 130925586 0 17976200 0 1112 0
cpu3 119675045 1978 158718307 839193898 130829935 0 18067354 0 2673 0
cpu4 119472865 2365 158335074 839980802 131284720 0 17839412 0 5737 0
cpu5 119645773 3445 158309726 839898352 131180071 0 17815874 0 2258 0
cpu6 121759510 3522 158645383 837163242 131276623 0 17887118 0 676 0
cpu7 119607143 1665 158323304 839899705 131279304 0 17739476 0 760 0
cpu8 119555754 1720 158826100 839307697 131192054 0 17920674 0 2963 0
cpu9 119331806 2430 158975846 839298371 131212030 0 17971455 0 1940 0
cpu10 119482412 2126 158776276 839387653 131245489 0 18026640 0 769 0
cpu11 119505370 1743 158953865 839216721 131042682 0 18003357 0 564 0
cpu12 124325811 9639 160042842 832380381 126636547 0 20963482 0 37521 0
cpu13 120160211 4258 160463299 836596324 131272789 0 18224669 0 10402 0
cpu14 120188434 3753 160355232 836895457 131194318 0 18215507 0 7435 0
cpu15 120194923 3498 160437708 836744702 131005045 0 18233329 0 56067 0
cpu16 120068858 2621 160064053 837608149 131245669 0 18087206 0 16442 0
cpu17 120069883 2506 159995819 837513363 131150376 0 18057910 0 1512 0
cpu18 119965182 2336 159919895 837704373 131253279 0 18083136 0 4330 0
cpu19 120041072 2161 159859469 837748115 131324420 0 18046819 0 5142 0
cpu20 120275965 2354 160356794 836189330 131653376 0 18261364 0 17004 0
cpu21 119901125 2449 160316176 837089702 131457348 0 18082668 0 1031 0
cpu22 119934355 2260 160462810 837142806 131289338 0 18138588 0 4630 0
cpu23 120055719 2574 160448305 837094662 131138804 0 18248811 0 3266 0
cpu24 68665235 2131 91563971 1002496028 102707445 0 8444683 0 1239 0
cpu25 68678065 583 91583992 1002706946 102719733 0 8405183 0 447 0
cpu26 68606442 1577 91572622 1002521776 102868164 0 8429542 0 496 0
cpu27 68610522 1080 91560315 1002763574 102772821 0 8418883 0 542 0
cpu28 68535817 2053 91253089 1003263517 102825277 0 8344881 0 524 0
cpu29 68543886 1754 91155120 1003344337 102820014 0 8318624 0 443 0
cpu30 68569440 1683 91204553 1003307014 102731992 0 8352947 0 702 0
cpu31 68501518 1598 91235747 1003291519 102920038 0 8311560 0 762 0
cpu32 68559777 1173 91507612 1002637991 102955265 0 8432026 0 1387 0
cpu33 68467407 2144 91646502 1002642064 103002650 0 8379092 0 216 0
cpu34 68479922 1526 91493457 1002801859 102950450 0 8411378 0 446 0
cpu35 68536233 1537 91575752 1002618159 102806054 0 8455260 0 535 0
cpu36 72006194 2763 93936392 995215166 101462754 0 9877716 0 2835 0
cpu37 69451747 2754 93126028 998648143 104183981 0 8607023 0 2732 0
cpu38 69476588 1461 93190060 998683209 104120446 0 8593668 0 130 0
cpu39 69416981 1522 93142610 998811813 104066518 0 8634348 0 34 0
cpu40 69350586 1692 92864186 999290306 104255841 0 8489781 0 2678 0
cpu41 69367416 1500 92681600 999449452 104166148 0 8498654 0 2894 0
cpu42 69392346 1730 92716154 999237028 104230296 0 8586659 0 297 0
cpu43 69401578 1014 92821956 999321748 104117726 0 8552287 0 56 0
cpu44 69385227 1329 93050556 998727766 104292239 0 8624834 0 44 0
cpu45 69343753 1805 93094116 998737869 104324994 0 8574365 0 66 0
cpu46 69375647 719 93111614 998677302 104406858 0 8557386 0 287 0
cpu47 69470150 2573 93173268 998377418 104364749 0 8653433 0 2936 0
intr 239337282674 52 0 0 0 0 0 0 0 1 28717933 0 0 0 0 0 0 0 0 231 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 254945358 249767277 327939120 313253960 59387928 65820234 29263911 27590397 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 386066789 637759344 795777142 821840471 1330315972 1300357630 1361616782 1310772194 169679810 62884035 329430665 219034668 3946862309 2411511044 2485096603 2531489889 3314242280 3176712852 3233371493 3271137768 2068393270 2077553042 2164525903 2130690865 2506559666 2658602766 2606580481 2571981869 2912649620 2813306522 2881072343 2835873325 2234940140 2145915612 2277940613 2293589188 2809139688 4040408136 4086330407 4065837554 109151820 46681509 112481469 133198980 3731816399 3792341041 3827141643 3823065888 1420 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 2 2 2 2 2 2 2 0 2 0 2 2 2 2 2 2 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
ctxt 6090442739525
btime 1458576335
processes 11265682
procs_running 1
procs_blocked 0
softirq 172496314827 116 4159040898 18511278 4189191998 1324881224 0 156282 490406286 0 3400336793
```

My longest line is significantly longer than yours (3756 vs 1021). I'm not sure that is the issue, but if you're able to reproduce on a larger machine it's worth looking into..

**Comment by IRCody** [comment]
Changing [this](https://github.com/intelsdi-x/snap-plugin-collector-cpu/blob/master/cpu/cpu.go#L302) line to, instead of grabbing the last character (which only works if cpu# single digit) to grab everything after cpu fixes the issue for me.

I'm still not sure:
* Why this causes the plugin to hang ever
* Why the plugin will successfully complete a few rounds of collection prior to hanging

I don't believe this is the full fix but maybe it will be helpful to figuring out the issue.

**Comment by katarzyna-z** [comment]
I've reproduced the issue on the machine with 70 core.

@IRCody After fixing the line which you mentioned the code in those lines [L366](https://github.com/intelsdi-x/snap-plugin-collector-cpu/blob/master/cpu/cpu.go#L366) and [L371](https://github.com/intelsdi-x/snap-plugin-collector-cpu/blob/master/cpu/cpu.go#L371) will not be executed. It seems that it explains why plugin is hung after a few collection (on my machine after 3 collection).


After changing L366 and L371 into `continue` without fixing the line which @IRCody mentioned the plugin works.

On the other hand it seems that change in the line which @IRCody mentioned is needed to fix the issue #28.  @IRCody Could you prepare the pull request with the fix for this issue?

This is what I know for now.

**Comment by IRCody** [comment]
I'm not sure how the code on those lines would be hanging though? That might be worth investigating since it's not obvious to me. I'm going to look into it more this week.

**Comment by katarzyna-z** [comment]
I've noticed that errors occurs when plugin sends logs to stderr, it means this code is executed:

```fmt.Fprintf(os.Stderr, "Percentage value of %v could not be calculated due to invalid data reported by /proc/stat", getNamespaceMetricPart(metricName, percentageRepresentationType))```

In the plugin's code, logging to stderr should appear only if percentage metrics cannot be calculated - after fix for the issue #28 this problem should disappear.

I've prepared the special versions of cpu plugin to observe what if plugin sends logs frequently:
- cpu plugin with frequent short log, the version is available [here](https://github.com/katarzyna-z/snap-plugin-collector-cpu/commit/a79c0a040d6308a9276a54d2e6f8cd27581f8b95), 
- cpu plugin with long log, the version is is available [here](https://github.com/katarzyna-z/snap-plugin-collector-cpu/commit/9738a2b2145067d19cb5e3c776ec9d233428b204) 

I've tested those two special versions of cpu plugin with two versions of snap:
- snap with capturing logs from plugins, it is this [commit]( https://github.com/intelsdi-x/snap/commit/5d91b04fa8b7d398fea04f5cbbe970f21cf32bd9),
- snap without capturing logs from plugins, it is this [commit](https://github.com/intelsdi-x/snap/commit/84aa75e8074ecbf633b43d07a96a12aa43f6861a).

I've observed that errors appear only in snap with capturing logs from plugins. It seems that there is a problem in capturing stderr from plugin when log line is long. In the snap version without capturing logs from plugins there was a special case for errors returned by the scanner, please see [this](https://github.com/intelsdi-x/snap/commit/5d91b04fa8b7d398fea04f5cbbe970f21cf32bd9#diff-9263d577284097ec1bec3e3fce13eb32L328).

Adding of new line at the end of log  which is sent to stderr fixes the issue on plugin side.

**Comment by katarzyna-z** [comment]
Fixed by pull request #35.
